go/vt/topotools/keyspace.go

```
@@ -0,0 +1,119 @@
+	default:
+		return err
+	}
+
+	// Any errors from this point onward are ignored.
+	var wg sync.WaitGroup
+	for _, ti := range tabletMap {
+		if ti.Hostname == "" {
+			// The tablet is not running, we don't have the host
+			// name to connect to, so we just skip this tablet.
+			logger.Infof("Tablet %v has no hostname, skipping its RefreshState", ti.AliasString())
+			continue
+		}
+
+		wg.Add(1)
+		go func(ti *topo.TabletInfo) {
+			defer wg.Done()
+			logger.Infof("Calling RefreshState on tablet %v", ti.AliasString())
+
+			ctx, cancel := context.WithTimeout(ctx, 60*time.Second)
+			defer cancel()
+
+			if err := tmc.RefreshState(ctx, ti.Tablet); err != nil {
+				logger.Warningf("RefreshTabletsByShard: failed to refresh %v: %v", ti.AliasString(), err)
+			}
+		}(ti)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+// UpdateShardRecords updates the shard records based on 'from' or 'to'
+// direction.
+func UpdateShardRecords(
+	ctx context.Context,
+	ts *topo.Server,
+	tmc tmclient.TabletManagerClient,
+	keyspace string,
+	shards []*topo.ShardInfo,
+	cells []string,
+	servedType topodatapb.TabletType,
+	isFrom bool,
+	clearSourceShards bool,
+	logger logutil.Logger,
+) error {
+	disableQueryService := isFrom
+	if err := ts.UpdateDisableQueryService(ctx, keyspace, shards, servedType, cells, disableQueryService); err != nil {
+		return err
+	}
+
+	for i, si := range shards {
+		updatedShard, err := ts.UpdateShardFields(ctx, si.Keyspace(), si.ShardName(), func(si *topo.ShardInfo) error {
+			if clearSourceShards {
+				si.SourceShards = nil
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			return err
+		}
+
+		shards[i] = updatedShard
+
+		// For 'to' shards, refresh to make them serve. The 'from' shards will
+		// be refreshed after traffic has migrated.
+		if !isFrom {
+			_ = RefreshTabletsByShard(ctx, ts, tmc, si, cells, logger)
```

I like this pattern of being explicit that we are ignoring the error. However, is not a pattern that I've seen in the codebase.  

I wonder if start adding here will create confusion.

---

My opinion is being that explicit about ignoring the error is a better (or at least more clear) pattern, and we've got to start somewhere! 😂  I'll do a quick temp check to see what some other folks think about this

---

That sounds like a plan. If we can get thumbs-up that we would like to make this a standard pattern moving forward, let's start here!

---

Consensus seemed to be:
* yes, explicitly acknowledging that we're ignoring errors is better than implicitly doing so
* we've had lots of trouble debugging tests specifically due to errors being discarded
* we should log the ignored error in this specific case, and whether to log or not is probably a case-by-case basis
* i'm also going to work on [adding `errcheck` to the linter](https://github.com/vitessio/vitess/pull/7996#issuecomment-829241274) so we can start to enforce this as a pattern going forward

---

Great, thanks for following up on this. All makes sense.